{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7717fb11",
   "metadata": {},
   "source": [
    "# Part 0: Load in libraries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97ea02f9",
   "metadata": {},
   "source": [
    "## runs using graph_env - needs d3blocks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9decc092",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/research/kiersten.ruff/fonts/ARIAL.TTF\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sbn\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import cm\n",
    "import re\n",
    "import math\n",
    "import scipy\n",
    "from scipy.stats import ks_2samp\n",
    "from matplotlib.patches import Rectangle\n",
    "import pathlib\n",
    "from matplotlib import font_manager\n",
    "from scipy.stats import fisher_exact\n",
    "\n",
    "\n",
    "# Setup the fonts\n",
    "fonts_path = pathlib.Path.home().joinpath('fonts') # i.e. `~/fonts` (update as needed)\n",
    "font_filename = 'ARIAL.TTF'\n",
    "font_path = fonts_path.joinpath(font_filename)\n",
    "print(font_path)\n",
    "\n",
    "font_manager.fontManager.addfont(str(font_path))\n",
    "prop = font_manager.FontProperties(fname=font_path)\n",
    "\n",
    "plt.rcParams['font.family'] = 'sans-serif'\n",
    "plt.rcParams['font.sans-serif'] = prop.get_name()\n",
    "plt.rcParams['pdf.fonttype'] = 42 # Makes text editiable\n",
    "plt.rcParams['ps.fonttype'] = 42\n",
    "\n",
    "COLOR_MAP = dict()\n",
    "COLOR_MAP['A']         = '#8c8c8c'\n",
    "COLOR_MAP['CHNQST']    = '#077a15'\n",
    "COLOR_MAP['DE']        = '#e23a36'\n",
    "COLOR_MAP['FWY']       = '#f58220'\n",
    "COLOR_MAP['G']         = '#86bd8d'\n",
    "COLOR_MAP['ILMV']      = '#231f20'\n",
    "COLOR_MAP['KR']        = '#1725c2'\n",
    "COLOR_MAP['P']         = '#c25fc9'\n",
    "\n",
    "\n",
    "# Using the global color map, we convert the colormap of amino acid\n",
    "# groups, to a map against individual amino acids. This\n",
    "# serves as our internal color map which is used for styling.\n",
    "aacolor_map=dict()\n",
    "for amino_acids in COLOR_MAP:\n",
    "    for amino_acid in amino_acids:\n",
    "        aacolor_map[amino_acid] = COLOR_MAP[amino_acids]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc4b254c",
   "metadata": {},
   "source": [
    "# Part 1: Connect spliceosome processes to IDR grammars"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03f0b4d6",
   "metadata": {},
   "source": [
    "## 1.1 - Load in spliceosome complex information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ba0df019",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Unnamed: 0 Accession     Gene    General Process  \\\n",
      "286         286    O00422    SAP18       ASAP complex   \n",
      "285         285    Q15287    RNPS1       ASAP complex   \n",
      "284         284    Q9UKV3    ACIN1       ASAP complex   \n",
      "0             0    Q13838   DDX39B  Major Spliceosome   \n",
      "170         170    Q15459    SF3A1  Major Spliceosome   \n",
      "..          ...       ...      ...                ...   \n",
      "365         365    Q9Y247   FAM50B   Splicing factors   \n",
      "366         366    Q8N1B3     CCNQ   Splicing factors   \n",
      "367         367    Q06787     FMR1   Splicing factors   \n",
      "401         401    O00425  IGF2BP3   Splicing factors   \n",
      "530         530    Q15695  ZRSR2P1   Splicing factors   \n",
      "\n",
      "                               Category  IDR Cluster 18 Cluster 26  \\\n",
      "286                        ASAP complex   No         No         No   \n",
      "285                        ASAP complex  Yes         No        Yes   \n",
      "284                        ASAP complex  Yes        Yes        Yes   \n",
      "0                Spliceosomal E complex   No         No         No   \n",
      "170  U2 small nuclear ribonucleoprotein  Yes         No         No   \n",
      "..                                  ...  ...        ...        ...   \n",
      "365     Other pre-mRNA splicing factors   No         No         No   \n",
      "366     Other pre-mRNA splicing factors   No         No         No   \n",
      "367     Other pre-mRNA splicing factors  Yes         No         No   \n",
      "401     Other pre-mRNA splicing factors   No         No         No   \n",
      "530     Other pre-mRNA splicing factors  Yes         No        Yes   \n",
      "\n",
      "    Cluster 18, or 26  \n",
      "286                No  \n",
      "285               Yes  \n",
      "284               Yes  \n",
      "0                  No  \n",
      "170                No  \n",
      "..                ...  \n",
      "365                No  \n",
      "366                No  \n",
      "367                No  \n",
      "401                No  \n",
      "530               Yes  \n",
      "\n",
      "[531 rows x 9 columns]\n",
      "['ASAP complex1', 'Major Spliceosome', 'Minor Spliceosome', 'RNA Modification', 'Other splicing factors1']\n",
      "['ASAP complex', 'Spliceosomal E complex', 'U2 small nuclear ribonucleoprotein', 'U1 small nuclear ribonucleoprotein', 'Spliceosomal P complex', 'NineTeen complex', 'tri-snRP complex', 'Exon junction complex', 'SF3b complex', 'LSm proteins', 'Sm spliceosomal proteins', 'NTC associated proteins', 'SF3a complex', 'U4/U6 small nuclear ribonucleoprotein particle', 'U5 small nuclear ribonucleoprotein', 'Spliceosomal B complex', 'Spliceosomal Bact complex', 'Spliceosomal A complex', 'Spliceosomal C complex', 'U11/U12 di-snRNP', 'WTAP complex', 'm6A methyltransferase complex', 'Other splicing factors', 'Serine and arginine rich splicing factors']\n",
      "24\n"
     ]
    }
   ],
   "source": [
    "dfcompclust=pd.read_excel('data/Nuclear_speckles_complexes_w_clusters_IDRs_gte_70_and_nonlinkers_gte_50.xlsx', engine='openpyxl')\n",
    "dfcompclust=dfcompclust.sort_values(by=['General Process'])\n",
    "print(dfcompclust)\n",
    "\n",
    "ccgeneralprocess=dfcompclust['General Process'].tolist()\n",
    "cccategory=dfcompclust['Category'].tolist()\n",
    "ccIDR=dfcompclust['IDR'].tolist()\n",
    "cccluster18=dfcompclust['Cluster 18'].tolist()\n",
    "cccluster26=dfcompclust['Cluster 26'].tolist()\n",
    "ccclusterany=dfcompclust['Cluster 18, or 26'].tolist()\n",
    "ccgene=dfcompclust['Gene'].tolist()\n",
    "ccacc=dfcompclust['Accession'].tolist()\n",
    "\n",
    "ccgeneralprocess = [x if x != 'Splicing factors' else 'Other splicing factors1' for x in ccgeneralprocess]\n",
    "ccgeneralprocess = [x if x != 'ASAP complex' else 'ASAP complex1' for x in ccgeneralprocess]\n",
    "cccategory = [x if x != 'Other pre-mRNA splicing factors' else 'Other splicing factors' for x in cccategory]\n",
    "\n",
    "uniccgeneralprocess=[]\n",
    "unicccategory=[]\n",
    "for s in range(0,len(ccgeneralprocess)):\n",
    "    if ccgeneralprocess[s] not in uniccgeneralprocess:\n",
    "        uniccgeneralprocess.append(ccgeneralprocess[s])\n",
    "    if cccategory[s] not in unicccategory:\n",
    "        unicccategory.append(cccategory[s])\n",
    "\n",
    "print(uniccgeneralprocess)\n",
    "print(unicccategory)\n",
    "print(len(unicccategory))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65adc97c",
   "metadata": {},
   "source": [
    "## 1.2 - Create data for plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "05c220af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4\n",
      "3\n",
      "2\n",
      "1\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "source=[]\n",
    "target=[]\n",
    "weight=[]\n",
    "\n",
    "for i in range(len(uniccgeneralprocess)-1,-1,-1): #range(0,len(uniccgeneralprocess)):\n",
    "    print(i)\n",
    "    for j in range(len(unicccategory)-1,-1,-1): #range(0,len(unicccategory)):\n",
    "        pos1=[m for m, x in enumerate(ccgeneralprocess) if x == uniccgeneralprocess[i]]\n",
    "        pos2=[m for m, x in enumerate(cccategory) if x == unicccategory[j]]\n",
    "        posinter=list(set(pos1) & set(pos2))\n",
    "\n",
    "\n",
    "        if posinter:\n",
    "            source.append(uniccgeneralprocess[i])\n",
    "            target.append(unicccategory[j])\n",
    "            weight.append(len(posinter))\n",
    "\n",
    "for j in range(len(unicccategory)-1,-1,-1): #range(0,len(unicccategory)):\n",
    "    pos2=[m for m, x in enumerate(cccategory) if x == unicccategory[j]]        \n",
    "    tmpidr=0\n",
    "    tmp18=0\n",
    "    tmp26=0\n",
    "    for p in pos2:\n",
    "        if ccIDR[p]=='Yes':\n",
    "            tmpidr=tmpidr+1\n",
    "        if cccluster18[p]=='Yes':\n",
    "            tmp18=tmp18+1\n",
    "        if cccluster26[p]=='Yes':\n",
    "            tmp26=tmp26+1\n",
    "\n",
    "    if tmpidr==0:\n",
    "        source.append(unicccategory[j])\n",
    "        target.append('No IDRs in Complex')\n",
    "        weight.append(1)\n",
    "\n",
    "    elif tmpidr>0 and tmp18+tmp26==0:\n",
    "        source.append(unicccategory[j])\n",
    "        target.append('IDRs in complex not 18 or 26')\n",
    "        weight.append(tmpidr)\n",
    "\n",
    "    elif tmpidr>0 and tmp18+tmp26>0:\n",
    "        if tmp18>0:\n",
    "            source.append(unicccategory[j])\n",
    "            target.append('IDRs in complex in 18')\n",
    "            weight.append(tmp18)\n",
    "        if tmp26>0:\n",
    "            source.append(unicccategory[j])\n",
    "            target.append('IDRs in complex in 26')\n",
    "            weight.append(tmp26)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "533df2e7",
   "metadata": {},
   "source": [
    "## 1.3 - Plot data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aa7a56c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                            source  \\\n",
      "0                          Other splicing factors1   \n",
      "1                          Other splicing factors1   \n",
      "2                                 RNA Modification   \n",
      "3                                 RNA Modification   \n",
      "4                                Minor Spliceosome   \n",
      "5                                Major Spliceosome   \n",
      "6                                Major Spliceosome   \n",
      "7                                Major Spliceosome   \n",
      "8                                Major Spliceosome   \n",
      "9                                Major Spliceosome   \n",
      "10                               Major Spliceosome   \n",
      "11                               Major Spliceosome   \n",
      "12                               Major Spliceosome   \n",
      "13                               Major Spliceosome   \n",
      "14                               Major Spliceosome   \n",
      "15                               Major Spliceosome   \n",
      "16                               Major Spliceosome   \n",
      "17                               Major Spliceosome   \n",
      "18                               Major Spliceosome   \n",
      "19                               Major Spliceosome   \n",
      "20                               Major Spliceosome   \n",
      "21                               Major Spliceosome   \n",
      "22                               Major Spliceosome   \n",
      "23                                   ASAP complex1   \n",
      "24       Serine and arginine rich splicing factors   \n",
      "25                          Other splicing factors   \n",
      "26                          Other splicing factors   \n",
      "27                   m6A methyltransferase complex   \n",
      "28                                    WTAP complex   \n",
      "29                                    WTAP complex   \n",
      "30                                U11/U12 di-snRNP   \n",
      "31                                U11/U12 di-snRNP   \n",
      "32                          Spliceosomal C complex   \n",
      "33                          Spliceosomal C complex   \n",
      "34                          Spliceosomal A complex   \n",
      "35                          Spliceosomal A complex   \n",
      "36                       Spliceosomal Bact complex   \n",
      "37                       Spliceosomal Bact complex   \n",
      "38                          Spliceosomal B complex   \n",
      "39                          Spliceosomal B complex   \n",
      "40              U5 small nuclear ribonucleoprotein   \n",
      "41              U5 small nuclear ribonucleoprotein   \n",
      "42  U4/U6 small nuclear ribonucleoprotein particle   \n",
      "43                                    SF3a complex   \n",
      "44                         NTC associated proteins   \n",
      "45                         NTC associated proteins   \n",
      "46                        Sm spliceosomal proteins   \n",
      "47                                    LSm proteins   \n",
      "48                                    SF3b complex   \n",
      "49                           Exon junction complex   \n",
      "50                                tri-snRP complex   \n",
      "51                                tri-snRP complex   \n",
      "52                                NineTeen complex   \n",
      "53                          Spliceosomal P complex   \n",
      "54                          Spliceosomal P complex   \n",
      "55              U1 small nuclear ribonucleoprotein   \n",
      "56              U2 small nuclear ribonucleoprotein   \n",
      "57                          Spliceosomal E complex   \n",
      "58                                    ASAP complex   \n",
      "59                                    ASAP complex   \n",
      "\n",
      "                                            target  weight  \n",
      "0        Serine and arginine rich splicing factors      12  \n",
      "1                           Other splicing factors     244  \n",
      "2                    m6A methyltransferase complex       4  \n",
      "3                                     WTAP complex       7  \n",
      "4                                 U11/U12 di-snRNP      10  \n",
      "5                           Spliceosomal C complex      51  \n",
      "6                           Spliceosomal A complex      19  \n",
      "7                        Spliceosomal Bact complex      27  \n",
      "8                           Spliceosomal B complex      19  \n",
      "9               U5 small nuclear ribonucleoprotein       8  \n",
      "10  U4/U6 small nuclear ribonucleoprotein particle       5  \n",
      "11                                    SF3a complex       3  \n",
      "12                         NTC associated proteins      16  \n",
      "13                        Sm spliceosomal proteins       8  \n",
      "14                                    LSm proteins      14  \n",
      "15                                    SF3b complex       7  \n",
      "16                           Exon junction complex       5  \n",
      "17                                tri-snRP complex       4  \n",
      "18                                NineTeen complex       6  \n",
      "19                          Spliceosomal P complex      47  \n",
      "20              U1 small nuclear ribonucleoprotein       3  \n",
      "21              U2 small nuclear ribonucleoprotein       5  \n",
      "22                          Spliceosomal E complex       4  \n",
      "23                                    ASAP complex       3  \n",
      "24                           IDRs in complex in 26      11  \n",
      "25                           IDRs in complex in 18      11  \n",
      "26                           IDRs in complex in 26      35  \n",
      "27                    IDRs in complex not 18 or 26       4  \n",
      "28                           IDRs in complex in 18       1  \n",
      "29                           IDRs in complex in 26       4  \n",
      "30                           IDRs in complex in 18       1  \n",
      "31                           IDRs in complex in 26       4  \n",
      "32                           IDRs in complex in 18       8  \n",
      "33                           IDRs in complex in 26       6  \n",
      "34                           IDRs in complex in 18       2  \n",
      "35                           IDRs in complex in 26       7  \n",
      "36                           IDRs in complex in 18       4  \n",
      "37                           IDRs in complex in 26       2  \n",
      "38                           IDRs in complex in 18       2  \n",
      "39                           IDRs in complex in 26       5  \n",
      "40                           IDRs in complex in 18       1  \n",
      "41                           IDRs in complex in 26       1  \n",
      "42                           IDRs in complex in 18       1  \n",
      "43                    IDRs in complex not 18 or 26       2  \n",
      "44                           IDRs in complex in 18       4  \n",
      "45                           IDRs in complex in 26       3  \n",
      "46                    IDRs in complex not 18 or 26       2  \n",
      "47                    IDRs in complex not 18 or 26       4  \n",
      "48                    IDRs in complex not 18 or 26       3  \n",
      "49                           IDRs in complex in 18       1  \n",
      "50                           IDRs in complex in 18       1  \n",
      "51                           IDRs in complex in 26       2  \n",
      "52                           IDRs in complex in 18       2  \n",
      "53                           IDRs in complex in 18       8  \n",
      "54                           IDRs in complex in 26       7  \n",
      "55                           IDRs in complex in 26       1  \n",
      "56                    IDRs in complex not 18 or 26       2  \n",
      "57                           IDRs in complex in 26       2  \n",
      "58                           IDRs in complex in 18       1  \n",
      "59                           IDRs in complex in 26       2  \n",
      "              source                 target  weight\n",
      "52  NineTeen complex  IDRs in complex in 18       2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[d3blocks] >INFO> Cleaning edge_properties and config parameters..\n",
      "[d3blocks] >INFO> Initializing [Sankey]\n",
      "[d3blocks] >INFO> filepath is set to [/tmp/d3blocks/sankey.html]\n",
      "[d3blocks] >INFO> Convert to DataFrame.\n",
      "[d3blocks] >INFO> Node properties are set.\n",
      "[d3blocks] >INFO> Edge properties are set.\n",
      "[d3blocks] >INFO> Set [figsize]: [500, 800]\n",
      "[d3blocks] >INFO> File already exists and will be overwritten: [/tmp/d3blocks/sankey.html]\n",
      "[d3blocks] >INFO> Open browser: /tmp/d3blocks/sankey.html\n"
     ]
    }
   ],
   "source": [
    "dftmp=pd.DataFrame()\n",
    "\n",
    "dftmp['source']=source\n",
    "dftmp['target']=target\n",
    "dftmp['weight']=weight\n",
    "\n",
    "print(dftmp)\n",
    "print(dftmp[dftmp['source']=='NineTeen complex'])\n",
    "\n",
    "\n",
    "from d3blocks import D3Blocks\n",
    "d3 = D3Blocks(chart='Sankey', frame=True)\n",
    "\n",
    "tmpcolorlist=['#3d2663', '#872f74', '#cb3f73', '#ff6361','#276026', '#008770', '#00aabe', '#64c8ff']\n",
    "\n",
    "\n",
    "#d3.set_node_properties(dftmp)\n",
    "d3.set_node_properties(dftmp, padding=1, color={'IDRs in complex in 18':'#000000', 'IDRs in complex in 26':'#1d9bd7','IDRs in complex not in 7, 18, or 23':'#dddddd','No IDRs in Complex':'#dddddd', \\\n",
    "                                     'Major Spliceosome':'#276026', 'Other splicing factors1':'#3d2663', 'RNA Modification':'#cb3f73', 'Minor Spliceosome':'#872f74', 'ASAP complex1':'#ff6361'})\n",
    "#d3.set_node_properties(dftmp, color=mycolorsdict)\n",
    "d3.set_edge_properties(dftmp, color='target', opacity='target')\n",
    "#d3.show(figsize=[600, 600])\n",
    "d3.show(figsize=[500, 800])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caec48dc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
