{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "21f3f485",
   "metadata": {},
   "source": [
    "# Part 0: Load in libraries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5549ec2",
   "metadata": {},
   "source": [
    "## runs using graph_env - needs d3blocks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "11988dc3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/research/kiersten.ruff/fonts/ARIAL.TTF\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sbn\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import cm\n",
    "import re\n",
    "import math\n",
    "import scipy\n",
    "from scipy.stats import ks_2samp\n",
    "from matplotlib.patches import Rectangle\n",
    "import pathlib\n",
    "from matplotlib import font_manager\n",
    "from scipy.stats import fisher_exact\n",
    "\n",
    "\n",
    "# Setup the fonts\n",
    "fonts_path = pathlib.Path.home().joinpath('fonts') # i.e. `~/fonts` (update as needed)\n",
    "font_filename = 'ARIAL.TTF'\n",
    "font_path = fonts_path.joinpath(font_filename)\n",
    "print(font_path)\n",
    "\n",
    "font_manager.fontManager.addfont(str(font_path))\n",
    "prop = font_manager.FontProperties(fname=font_path)\n",
    "\n",
    "plt.rcParams['font.family'] = 'sans-serif'\n",
    "plt.rcParams['font.sans-serif'] = prop.get_name()\n",
    "plt.rcParams['pdf.fonttype'] = 42 # Makes text editiable\n",
    "plt.rcParams['ps.fonttype'] = 42\n",
    "\n",
    "COLOR_MAP = dict()\n",
    "COLOR_MAP['A']         = '#8c8c8c'\n",
    "COLOR_MAP['CHNQST']    = '#077a15'\n",
    "COLOR_MAP['DE']        = '#e23a36'\n",
    "COLOR_MAP['FWY']       = '#f58220'\n",
    "COLOR_MAP['G']         = '#86bd8d'\n",
    "COLOR_MAP['ILMV']      = '#231f20'\n",
    "COLOR_MAP['KR']        = '#1725c2'\n",
    "COLOR_MAP['P']         = '#c25fc9'\n",
    "\n",
    "\n",
    "# Using the global color map, we convert the colormap of amino acid\n",
    "# groups, to a map against individual amino acids. This\n",
    "# serves as our internal color map which is used for styling.\n",
    "aacolor_map=dict()\n",
    "for amino_acids in COLOR_MAP:\n",
    "    for amino_acid in amino_acids:\n",
    "        aacolor_map[amino_acid] = COLOR_MAP[amino_acids]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "447a01c6",
   "metadata": {},
   "source": [
    "# Part 1: Connect nucleoplasm complexes to IDR grammars"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6be53b22",
   "metadata": {},
   "source": [
    "## 1.1 - Load in nucleoli complex information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "920b1466",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Unnamed: 0 Accession     Gene                           General Process  \\\n",
      "84           84    P51532  SMARCA4              Chromatin remodeling complex   \n",
      "76           76    O94805   ACTL6B              Chromatin remodeling complex   \n",
      "75           75    O96019   ACTL6A              Chromatin remodeling complex   \n",
      "74           74    P60709     ACTB              Chromatin remodeling complex   \n",
      "73           73    Q969G3  SMARCE1              Chromatin remodeling complex   \n",
      "..          ...       ...      ...                                       ...   \n",
      "166         166    Q9NPA8     ENY2  RNA polymerase II-mediated transcription   \n",
      "167         167    Q92830    KAT2A  RNA polymerase II-mediated transcription   \n",
      "169         169    Q96ES7    SGF29  RNA polymerase II-mediated transcription   \n",
      "157         157    P53803   POLR2K  RNA polymerase II-mediated transcription   \n",
      "0             0    P24863     CCNC  RNA polymerase II-mediated transcription   \n",
      "\n",
      "                       Category  \\\n",
      "84                ncBAF complex   \n",
      "76                ncBAF complex   \n",
      "75                ncBAF complex   \n",
      "74                ncBAF complex   \n",
      "73                 pBAF complex   \n",
      "..                          ...   \n",
      "166                SAGA complex   \n",
      "167                SAGA complex   \n",
      "169                SAGA complex   \n",
      "157  RNA Polymerase II subunits   \n",
      "0              Mediator complex   \n",
      "\n",
      "                                              Location  Cluster Number Test  \\\n",
      "84   Nucleoli, Nucleoli fibrillar center, Nucleoli ...                  7.0   \n",
      "76                                                 NaN                  NaN   \n",
      "75                                                 NaN                  NaN   \n",
      "74                                                 NaN                  NaN   \n",
      "73                                         Nucleoplasm                  7.0   \n",
      "..                                                 ...                  ...   \n",
      "166                                                NaN                  NaN   \n",
      "167                                                  0                  6.0   \n",
      "169                                                NaN                  NaN   \n",
      "157                                                NaN                  NaN   \n",
      "0                                          Nucleoplasm                  1.0   \n",
      "\n",
      "                                               Source  IDR Cluster 10  \\\n",
      "84      https://doi.org/10.1016/j.molcel.2023.03.013   Yes        Yes   \n",
      "76      https://doi.org/10.1016/j.molcel.2023.03.013    No         No   \n",
      "75      https://doi.org/10.1016/j.molcel.2023.03.013    No         No   \n",
      "74      https://doi.org/10.1016/j.molcel.2023.03.013    No         No   \n",
      "73      https://doi.org/10.1016/j.molcel.2023.03.013   Yes         No   \n",
      "..                                                ...  ...        ...   \n",
      "166  https://doi.org/10.1016/j.tibs.2017.09.001, HGNC   No         No   \n",
      "167  https://doi.org/10.1016/j.tibs.2017.09.001, HGNC  Yes         No   \n",
      "169  https://doi.org/10.1016/j.tibs.2017.09.001, HGNC   No         No   \n",
      "157   https://doi.org/10.3389/fmolb.2021.680090, HGNC   No         No   \n",
      "0                                                HGNC   No         No   \n",
      "\n",
      "    Cluster 11 Cluster 28 Cluster 10, 11, or 28  \n",
      "84          No         No                   Yes  \n",
      "76          No         No                    No  \n",
      "75          No         No                    No  \n",
      "74          No         No                    No  \n",
      "73          No         No                    No  \n",
      "..         ...        ...                   ...  \n",
      "166         No         No                    No  \n",
      "167         No         No                    No  \n",
      "169         No         No                    No  \n",
      "157         No         No                    No  \n",
      "0           No         No                    No  \n",
      "\n",
      "[487 rows x 13 columns]\n",
      "['Chromatin remodeling complex', 'Enhanceosomes', 'Histone modifying complex', 'RNA binding', 'RNA polymerase II-mediated transcription']\n",
      "['ncBAF complex', 'pBAF complex', 'BAF complex', 'INO80', 'ISWI complexes', 'CHD complexes', 'Wnt enhanceosome complex', 'Interferon-beta enhanceosome complex', 'Tip60/Nua4 histone acetyltransferase complex', 'ATAC complex', 'COMPASS-like complexes', 'p160 steroid receptor coactivator (SRC) family', 'RNA binding motif containing (RBM)', 'PIC', 'Mediator complex', 'SAGA complex', 'Super elongation complex', 'CREBBP/EP300', 'RNA Polymerase II subunits']\n",
      "19\n"
     ]
    }
   ],
   "source": [
    "dfcompclust=pd.read_excel('data/Nucleoplasm_processes_w_clusters_IDRs_gte_70_and_nonlinkers_gte_50.xlsx', engine='openpyxl')\n",
    "dfcompclust=dfcompclust.sort_values(by=['General Process'])\n",
    "print(dfcompclust)\n",
    "\n",
    "ccgeneralprocess=dfcompclust['General Process'].tolist()\n",
    "cccategory=dfcompclust['Category'].tolist()\n",
    "ccIDR=dfcompclust['IDR'].tolist()\n",
    "cccluster10=dfcompclust['Cluster 10'].tolist()\n",
    "cccluster11=dfcompclust['Cluster 11'].tolist()\n",
    "cccluster28=dfcompclust['Cluster 28'].tolist()\n",
    "ccclusterany=dfcompclust['Cluster 10, 11, or 28'].tolist()\n",
    "ccgene=dfcompclust['Gene'].tolist()\n",
    "ccacc=dfcompclust['Accession'].tolist()\n",
    "\n",
    "ccgeneralprocess = [x if x != 'Splicing factors' else 'Other splicing factors1' for x in ccgeneralprocess]\n",
    "ccgeneralprocess = [x if x != 'ASAP complex' else 'ASAP complex1' for x in ccgeneralprocess]\n",
    "cccategory = [x if x != 'CREBBP/EP301' else 'CREBBP/EP300' for x in cccategory]\n",
    "\n",
    "uniccgeneralprocess=[]\n",
    "unicccategory=[]\n",
    "for s in range(0,len(ccgeneralprocess)):\n",
    "    if ccgeneralprocess[s] not in uniccgeneralprocess:\n",
    "        uniccgeneralprocess.append(ccgeneralprocess[s])\n",
    "    if cccategory[s] not in unicccategory:\n",
    "        unicccategory.append(cccategory[s])\n",
    "\n",
    "print(uniccgeneralprocess)\n",
    "print(unicccategory)\n",
    "print(len(unicccategory))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a577ebcd",
   "metadata": {},
   "source": [
    "## 1.2 - Create data for plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a80c10b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4\n",
      "3\n",
      "2\n",
      "1\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "source=[]\n",
    "target=[]\n",
    "weight=[]\n",
    "\n",
    "for i in range(len(uniccgeneralprocess)-1,-1,-1): #range(0,len(uniccgeneralprocess)):\n",
    "    print(i)\n",
    "    for j in range(len(unicccategory)-1,-1,-1): #range(0,len(unicccategory)):\n",
    "        pos1=[m for m, x in enumerate(ccgeneralprocess) if x == uniccgeneralprocess[i]]\n",
    "        pos2=[m for m, x in enumerate(cccategory) if x == unicccategory[j]]\n",
    "        posinter=list(set(pos1) & set(pos2))\n",
    "\n",
    "\n",
    "        if posinter:\n",
    "            source.append(uniccgeneralprocess[i])\n",
    "            target.append(unicccategory[j])\n",
    "            weight.append(len(posinter))\n",
    "\n",
    "for j in range(len(unicccategory)-1,-1,-1): #range(0,len(unicccategory)):\n",
    "    pos2=[m for m, x in enumerate(cccategory) if x == unicccategory[j]]        \n",
    "    tmpidr=0\n",
    "    tmp10=0\n",
    "    tmp11=0\n",
    "    tmp28=0\n",
    "    for p in pos2:\n",
    "        if ccIDR[p]=='Yes':\n",
    "            tmpidr=tmpidr+1\n",
    "        if cccluster10[p]=='Yes':\n",
    "            tmp10=tmp10+1\n",
    "        if cccluster11[p]=='Yes':\n",
    "            tmp11=tmp11+1\n",
    "        if cccluster28[p]=='Yes':\n",
    "            tmp28=tmp28+1\n",
    "\n",
    "    if tmpidr==0:\n",
    "        source.append(unicccategory[j])\n",
    "        target.append('No IDRs in Complex')\n",
    "        weight.append(1)\n",
    "\n",
    "    elif tmpidr>0 and tmp10+tmp11+tmp28==0:\n",
    "        source.append(unicccategory[j])\n",
    "        target.append('IDRs in complex not in 10, 11, or 28')\n",
    "        weight.append(tmpidr)\n",
    "\n",
    "    elif tmpidr>0 and tmp10+tmp11+tmp28>0:\n",
    "        if tmp11>0:\n",
    "            source.append(unicccategory[j])\n",
    "            target.append('IDRs in complex in 11')\n",
    "            weight.append(tmp11)\n",
    "        if tmp28>0:\n",
    "            source.append(unicccategory[j])\n",
    "            target.append('IDRs in complex in 28')\n",
    "            weight.append(tmp28)\n",
    "        if tmp10>0:\n",
    "            source.append(unicccategory[j])\n",
    "            target.append('IDRs in complex in 10')\n",
    "            weight.append(tmp10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aca93f5b",
   "metadata": {},
   "source": [
    "## 1.3 - Plot data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e4e2e01b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[d3blocks] >INFO> Cleaning edge_properties and config parameters..\n",
      "[d3blocks] >INFO> Initializing [Sankey]\n",
      "[d3blocks] >INFO> filepath is set to [/tmp/d3blocks/sankey.html]\n",
      "[d3blocks] >INFO> Convert to DataFrame.\n",
      "[d3blocks] >INFO> Node properties are set.\n",
      "[d3blocks] >INFO> Edge properties are set.\n",
      "[d3blocks] >INFO> Set [figsize]: [500, 800]\n",
      "[d3blocks] >INFO> File already exists and will be overwritten: [/tmp/d3blocks/sankey.html]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                            source  \\\n",
      "0         RNA polymerase II-mediated transcription   \n",
      "1         RNA polymerase II-mediated transcription   \n",
      "2         RNA polymerase II-mediated transcription   \n",
      "3         RNA polymerase II-mediated transcription   \n",
      "4         RNA polymerase II-mediated transcription   \n",
      "5         RNA polymerase II-mediated transcription   \n",
      "6                                      RNA binding   \n",
      "7                        Histone modifying complex   \n",
      "8                        Histone modifying complex   \n",
      "9                        Histone modifying complex   \n",
      "10                       Histone modifying complex   \n",
      "11                                   Enhanceosomes   \n",
      "12                                   Enhanceosomes   \n",
      "13                    Chromatin remodeling complex   \n",
      "14                    Chromatin remodeling complex   \n",
      "15                    Chromatin remodeling complex   \n",
      "16                    Chromatin remodeling complex   \n",
      "17                    Chromatin remodeling complex   \n",
      "18                    Chromatin remodeling complex   \n",
      "19                      RNA Polymerase II subunits   \n",
      "20                                    CREBBP/EP300   \n",
      "21                                    CREBBP/EP300   \n",
      "22                        Super elongation complex   \n",
      "23                                    SAGA complex   \n",
      "24                                Mediator complex   \n",
      "25                                Mediator complex   \n",
      "26                                Mediator complex   \n",
      "27                                             PIC   \n",
      "28              RNA binding motif containing (RBM)   \n",
      "29              RNA binding motif containing (RBM)   \n",
      "30              RNA binding motif containing (RBM)   \n",
      "31  p160 steroid receptor coactivator (SRC) family   \n",
      "32  p160 steroid receptor coactivator (SRC) family   \n",
      "33                          COMPASS-like complexes   \n",
      "34                          COMPASS-like complexes   \n",
      "35                          COMPASS-like complexes   \n",
      "36                                    ATAC complex   \n",
      "37    Tip60/Nua4 histone acetyltransferase complex   \n",
      "38            Interferon-beta enhanceosome complex   \n",
      "39                        Wnt enhanceosome complex   \n",
      "40                        Wnt enhanceosome complex   \n",
      "41                                   CHD complexes   \n",
      "42                                   CHD complexes   \n",
      "43                                   CHD complexes   \n",
      "44                                  ISWI complexes   \n",
      "45                                           INO80   \n",
      "46                                     BAF complex   \n",
      "47                                     BAF complex   \n",
      "48                                     BAF complex   \n",
      "49                                    pBAF complex   \n",
      "50                                    pBAF complex   \n",
      "51                                   ncBAF complex   \n",
      "52                                   ncBAF complex   \n",
      "53                                   ncBAF complex   \n",
      "\n",
      "                                            target  weight  \n",
      "0                       RNA Polymerase II subunits      12  \n",
      "1                                     CREBBP/EP300       2  \n",
      "2                         Super elongation complex      17  \n",
      "3                                     SAGA complex      17  \n",
      "4                                 Mediator complex      33  \n",
      "5                                              PIC      29  \n",
      "6               RNA binding motif containing (RBM)     213  \n",
      "7   p160 steroid receptor coactivator (SRC) family       3  \n",
      "8                           COMPASS-like complexes      18  \n",
      "9                                     ATAC complex      13  \n",
      "10    Tip60/Nua4 histone acetyltransferase complex      18  \n",
      "11            Interferon-beta enhanceosome complex       6  \n",
      "12                        Wnt enhanceosome complex      18  \n",
      "13                                   CHD complexes       6  \n",
      "14                                  ISWI complexes      11  \n",
      "15                                           INO80      15  \n",
      "16                                     BAF complex      22  \n",
      "17                                    pBAF complex      19  \n",
      "18                                   ncBAF complex      15  \n",
      "19                           IDRs in complex in 28       1  \n",
      "20                           IDRs in complex in 11       2  \n",
      "21                           IDRs in complex in 10       2  \n",
      "22                           IDRs in complex in 10       1  \n",
      "23            IDRs in complex not in 10, 11, or 28      10  \n",
      "24                           IDRs in complex in 11       4  \n",
      "25                           IDRs in complex in 28       2  \n",
      "26                           IDRs in complex in 10       4  \n",
      "27                           IDRs in complex in 11       1  \n",
      "28                           IDRs in complex in 11       2  \n",
      "29                           IDRs in complex in 28      32  \n",
      "30                           IDRs in complex in 10       3  \n",
      "31                           IDRs in complex in 11       1  \n",
      "32                           IDRs in complex in 10       3  \n",
      "33                           IDRs in complex in 11       4  \n",
      "34                           IDRs in complex in 28       1  \n",
      "35                           IDRs in complex in 10       4  \n",
      "36            IDRs in complex not in 10, 11, or 28       8  \n",
      "37                           IDRs in complex in 11       1  \n",
      "38                           IDRs in complex in 10       1  \n",
      "39                           IDRs in complex in 28       1  \n",
      "40                           IDRs in complex in 10       6  \n",
      "41                           IDRs in complex in 11       1  \n",
      "42                           IDRs in complex in 28       2  \n",
      "43                           IDRs in complex in 10       1  \n",
      "44                           IDRs in complex in 11       1  \n",
      "45                           IDRs in complex in 10       1  \n",
      "46                           IDRs in complex in 11       4  \n",
      "47                           IDRs in complex in 28       2  \n",
      "48                           IDRs in complex in 10       5  \n",
      "49                           IDRs in complex in 11       4  \n",
      "50                           IDRs in complex in 10       3  \n",
      "51                           IDRs in complex in 11       2  \n",
      "52                           IDRs in complex in 28       2  \n",
      "53                           IDRs in complex in 10       3  \n",
      "Empty DataFrame\n",
      "Columns: [source, target, weight]\n",
      "Index: []\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[d3blocks] >INFO> Open browser: /tmp/d3blocks/sankey.html\n"
     ]
    }
   ],
   "source": [
    "dftmp=pd.DataFrame()\n",
    "\n",
    "dftmp['source']=source\n",
    "dftmp['target']=target\n",
    "dftmp['weight']=weight\n",
    "\n",
    "print(dftmp)\n",
    "print(dftmp[dftmp['source']=='NineTeen complex'])\n",
    "\n",
    "\n",
    "from d3blocks import D3Blocks\n",
    "d3 = D3Blocks(chart='Sankey', frame=True)\n",
    "\n",
    "tmpcolorlist=['#3d2663', '#872f74', '#cb3f73', '#ff6361','#276026', '#008770', '#00aabe', '#64c8ff']\n",
    "\n",
    "\n",
    "#d3.set_node_properties(dftmp)\n",
    "d3.set_node_properties(dftmp, padding=1, color={'IDRs in complex in 10':'#1d9bd7','IDRs in complex in 11':'#218342', 'IDRs in complex in 28':'#f58220','IDRs in complex not in 10, 11, or 28':'#dddddd','No IDRs in Complex':'#dddddd', \\\n",
    "                                     'RNA binding':'#276026', 'Enhanceosomes':'#3d2663', 'Chromatin remodeling complex':'#cb3f73', 'Histone modifying complex':'#872f74', 'RNA polymerase II-mediated transcription':'#ff6361'})\n",
    "#d3.set_node_properties(dftmp, color=mycolorsdict)\n",
    "d3.set_edge_properties(dftmp, color='target', opacity='target')\n",
    "#d3.show(figsize=[600, 600])\n",
    "d3.show(figsize=[500, 800])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00a59a28",
   "metadata": {},
   "source": [
    "## 1.4 - Save data to excel file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6097ec45",
   "metadata": {},
   "outputs": [],
   "source": [
    "with pd.ExcelWriter('../Supplementary_Tables/Table_S5.xlsx', engine='openpyxl',mode='a') as writer:\n",
    "    dftmp.to_excel(writer,sheet_name='Fig_5D')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6b915ca",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
